File: android/app/src/main/java/it/airgap/wallet/plugin/isolatedmodules/js/JSEvaluator.kt
package it.airgap.wallet.plugin.isolatedmodules.js

import android.content.Context
import android.os.Build
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import it.airgap.wallet.plugin.isolatedmodules.FileExplorer
import it.airgap.wallet.plugin.isolatedmodules.js.environment.JSEnvironment
import it.airgap.wallet.plugin.isolatedmodules.js.environment.JavaScriptEngineEnvironment
import it.airgap.wallet.plugin.isolatedmodules.js.environment.WebViewEnvironment
import it.airgap.wallet.util.asyncMap
import kotlinx.coroutines.*
import java.util.*

class JSEvaluator constructor(
    private val defaultEnvironment: JSEnvironment,
    private val environments: Map<JSEnvironment.Type, JSEnvironment?>
) {
    private val modules: MutableMap<String, JSModule> = mutableMapOf()

    fun registerModule(module: JSModule, protocolIdentifiers: List<String>) {
        module.registerFor(protocolIdentifiers)
    }

    fun deregisterModules(identifiers: List<String>) {
        identifiers.forEach { modules.remove(it) }
    }

    fun deregisterAllModules() {
        modules.clear()
    }

    suspend inline fun <T> singleRun(block: (runRef: String) -> T): T {
        val runRef = UUID.randomUUID().toString()
        return block(runRef).also { reset(runRef) }
    }

    suspend fun evaluatePreviewModule(module: JSModule, runRef: String? = null): JSObject =
        module.run(JSModuleAction.Load(null, null), runRef).also {
            module.appendType(it)
        }

    suspend fun evaluateLoadModules(
        modules: List<JSModule>,
        protocolType: JSProtocolType?,
        ignoreProtocols: JSArray?,
        runRef: String? = null
    ): JSObject {
        val modulesJson = modules.asyncMap { module ->
            module.run(JSModuleAction.Load(protocolType, ignoreProtocols), runRef).also {
                module.appendType(it)
                module.registerFor(it)
            }
        }

        return JSObject("""
            {
                "modules": $modulesJson
            }
        """.trimIndent())
    }

    suspend fun evaluateCallOfflineProtocolMethod(
        name: String,
        args: JSArray?,
        protocolIdentifier: String,
        runRef: String? = null
    ): JSObject {
        val module = modules[protocolIdentifier] ?: failWithModuleForProtocolNotFound(protocolIdentifier)
        return module.run(JSModuleAction.CallMethod.OfflineProtocol(name, args, protocolIdentifier), runRef)
    }

    suspend fun evaluateCallOnlineProtocolMethod(
        name: String,
        args: JSArray?,
        protocolIdentifier: String,
        networkId: String?,
        runRef: String? = null
    ): JSObject {
        val module = modules[protocolIdentifier] ?: failWithModuleForProtocolNotFound(protocolIdentifier)
        return module.run(JSModuleAction.CallMethod.OnlineProtocol(name, args, protocolIdentifier, networkId), runRef)
    }

    suspend fun evaluateCallBlockExplorerMethod(
        name: String,
        args: JSArray?,
        protocolIdentifier: String,
        networkId: String?,
        runRef: String? = null
    ): JSObject {
        val module = modules[protocolIdentifier] ?: failWithModuleForProtocolNotFound(protocolIdentifier)
        return module.run(JSModuleAction.CallMethod.BlockExplorer(name, args, protocolIdentifier, networkId), runRef)
    }

    suspend fun evaluateCallV3SerializerCompanionMethod(
        name: String,
        args: JSArray?,
        moduleIdentifier: String,
        runRef: String? = null
    ): JSObject {
        val module = modules[moduleIdentifier] ?: failWithModuleNotFound(moduleIdentifier)
        return module.run(JSModuleAction.CallMethod.V3SerializerCompanion(name, args), runRef)
    }

    suspend fun reset(runRef: String) {
        environments.values.forEach { it?.reset(runRef) }
    }

    suspend fun destroy() {
        environments.values.forEach { it?.destroy() }
    }

    private suspend fun JSModule.run(action: JSModuleAction, runRef: String? = null): JSObject {
        val environment = environments[preferredEnvironment] ?: defaultEnvironment

        return environment.run(this, action, runRef)
    }

    private fun JSModule.appendType(json: JSObject) {
        val type = when (this) {
            is JSModule.Asset -> "static"
            is JSModule.Installed, is JSModule.Preview -> "dynamic"
        }

        json.put("type", type)
    }

    private fun JSModule.registerFor(json: JSObject) {
        val protocols = json.getJSONArray("protocols")
        val protocolIdentifiers = buildList {
            for (i in 0 until protocols.length()) {
                val protocol = protocols.getJSONObject(i)
                add(protocol.getString("identifier"))
            }
        }

        registerFor(protocolIdentifiers)
    }

    private fun JSModule.registerFor(protocolIdentifiers: List<String>) {
        modules[identifier] = this
        protocolIdentifiers.forEach { modules[it] = this }
    }

    @Throws(IllegalStateException::class)
    private fun failWithModuleForProtocolNotFound(protocolIdentifier: String): Nothing = throw IllegalStateException("Module for protocol $protocolIdentifier could not be found.")

    @Throws(IllegalStateException::class)
    private fun failWithModuleNotFound(moduleIdentifier: String): Nothing = throw IllegalStateException("Module $moduleIdentifier could not be found.")
}

suspend fun JSEvaluator(context: Context, fileExplorer: FileExplorer): JSEvaluator {
    val webViewEnvironment = WebViewEnvironment(context, fileExplorer)
    val javaScriptEngineEnvironment =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) JavaScriptEngineEnvironment(context, fileExplorer).takeIf { it.isSupported() }
        else null

    val environments = mapOf(
        JSEnvironment.Type.WebView to webViewEnvironment,
        JSEnvironment.Type.JavaScriptEngine to javaScriptEngineEnvironment,
    )

    return JSEvaluator(webViewEnvironment, environments)
}

File: ios/App/App/IsolatedModules/JS/JSEvaluator.swift
//
//  JSEvaluator.swift
//  App
//
//  Created by Julia Samol on 06.02.23.
//

import Foundation
import Capacitor
import WebKit

class JSEvaluator {
    private let webViewEnv: WebViewEnvironment
    private let modulesManager: ModulesManager
    
    init(fileExplorer: FileExplorer) {
        self.webViewEnv = .init(fileExplorer: fileExplorer)
        self.modulesManager = .init()
    }
    
    func registerModule(_ module: JSModule, forProtocols protocolIdentifiers: [String]) async {
        await modulesManager.registerModule(module, forProtocols: protocolIdentifiers)
    }
    
    func deregisterModules(_ identifiers: [String]) async {
        await modulesManager.deregisterModules(identifiers)
    }
    
    func deregisterAllModules() async {
        await modulesManager.deregisterAllModules()
    }
    
    func singleRun<T>(_ block: (_ runRef: String) async throws -> T) async throws -> T {
        let runRef = UUID().uuidString
        let result = try await block(runRef)
        try await reset(runRef: runRef)
        
        return result
    }
    
    func evaluatePreviewModule(
        _ module: JSModule,
        ignoringProtocols ignoreProtocols: [String]?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        return try await self.webViewEnv.run(
            .load(.init(protocolType: nil, ignoreProtocols: ignoreProtocols)),
            in: module,
            ref: runRef
        )
    }
    
    func evaluateLoadModules(
        _ modules: [JSModule],
        for protocolType: JSProtocolType?,
        ignoringProtocols ignoreProtocols: [String]?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        let modulesJSON = try await modules.asyncMap { module -> [String : Any] in
            var json = try await self.webViewEnv.run(
                .load(.init(protocolType: protocolType, ignoreProtocols: ignoreProtocols)),
                in: module,
                ref: runRef
            )
            
            json["type"] = {
                switch module {
                case .asset(_):
                    return "static"
                case .installed(_), .preview(_):
                    return "dynamic"
                }
            }()
            
            try await self.modulesManager.registerModule(module, forJSON: json)
            
            return json
        }
        
        return ["modules": modulesJSON]
    }
    
    func evaluateCallOfflineProtocolMethod(
        _ name: String,
        ofProtocol protocolIdentifier: String,
        withArgs args: JSArray?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        let modules = await modulesManager.modules
        guard let module = modules[protocolIdentifier] else {
            throw Error.moduleNotFound(protocolIdentifier)
        }
        
        return try await webViewEnv.run(
            .callMethod(
                .offlineProtocol(
                    .init(name: name, args: args, protocolIdentifier: protocolIdentifier)
                )
            ),
            in: module,
            ref: runRef
        )
    }
    
    func evaluateCallOnlineProtocolMethod(
        _ name: String,
        ofProtocol protocolIdentifier: String,
        onNetwork networkID: String?,
        withArgs args: JSArray?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        let modules = await modulesManager.modules
        guard let module = modules[protocolIdentifier] else {
            throw Error.moduleNotFound(protocolIdentifier)
        }
        
        return try await webViewEnv.run(
            .callMethod(
                .onlineProtocol(
                    .init(name: name, args: args, protocolIdentifier: protocolIdentifier, networkID: networkID)
                )
            ),
            in: module,
            ref: runRef
        )
    }
    
    func evaluateCallBlockExplorerMethod(
        _ name: String,
        ofProtocol protocolIdentifier: String,
        onNetwork networkID: String?,
        withArgs args: JSArray?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        let modules = await modulesManager.modules
        guard let module = modules[protocolIdentifier] else {
            throw Error.moduleNotFound(protocolIdentifier)
        }
        
        return try await webViewEnv.run(
            .callMethod(
                .blockExplorer(
                    .init(name: name, args: args, protocolIdentifier: protocolIdentifier, networkID: networkID)
                )
            ),
            in: module,
            ref: runRef
        )
    }
    
    func evaluateCallV3SerializerCompanionMethod(
        _ name: String,
        ofModule moduleIdentifier: String,
        withArgs args: JSArray?,
        runRef: String? = nil
    ) async throws -> [String: Any] {
        let modules = await modulesManager.modules
        guard let module = modules[moduleIdentifier] else {
            throw Error.moduleNotFound(moduleIdentifier)
        }
        
        return try await webViewEnv.run(
            .callMethod(
                .v3SerializerCompanion(
                    .init(name: name, args: args)
                )
            ),
            in: module,
            ref: runRef
        )
    }
    
    func reset(runRef: String) async throws {
        try await webViewEnv.reset(runRef: runRef)
    }
    
    func destroy() async throws {
        try await webViewEnv.destroy()
    }
    
    private actor ModulesManager {
        private(set) var modules: [String: JSModule] = [:]
        
        func registerModule(_ module: JSModule, forJSON json: [String: Any]) throws {
            guard let protocols = json["protocols"] as? [Any] else {
                throw Error.invalidJSON
            }
            
            let protocolIdentifiers = try protocols.map { `protocol` -> String in
                guard let `protocol` = `protocol` as? [String: Any], let identifier = `protocol`["identifier"] as? String else {
                    throw Error.invalidJSON
                }
                
                return identifier
            }
            
            registerModule(module, forProtocols: protocolIdentifiers)
        }
        
        func registerModule(_ module: JSModule, forProtocols protocolIdentifiers: [String]) {
            modules[module.identifier] = module
            protocolIdentifiers.forEach { identifier in modules[identifier] = module }
        }
        
        func deregisterModules(_ identifiers: [String]) {
            identifiers.forEach { modules.removeValue(forKey: $0) }
        }
        
        func deregisterAllModules() {
            modules.removeAll()
        }
    }
    
    enum Error: Swift.Error {
        case moduleNotFound(String)
        case invalidJSON
    }
}
